example_buffer.py

```diff
@@ -0,0 +1,194 @@
+import tensorflow as tf
+import os
+import random
+from tqdm import tqdm
+import multiprocessing as mp
+import functools
+import itertools
+import rl_loop
+import dual_net
+import subprocess
+from utils import timer
+import time
+import preprocessing
+import argh
+import argparse
+import datetime as dt
+
+
+# 1 for ZLIB!  Make this match preprocessing.
+READ_OPTS = tf.python_io.TFRecordOptions(1)
+
+LOCAL_DIR = "data/"
+
+
+def pick_examples_from_tfrecord(filename, samples_per_game=4):
+    protos = [p for p in
+              tf.python_io.tf_record_iterator(filename, READ_OPTS)]
+    if len(protos) < 20:
+        return []
+    choices = random.sample(protos, samples_per_game)
+
+    def make_example(protostring):
+        e = tf.train.Example()
+        e.ParseFromString(protostring)
+        return e
+    examples = list(map(make_example, choices))
+    return examples
+
+
+def choose(g, samples_per_game=4):
+    examples = pick_examples_from_tfrecord(g, samples_per_game)
+    t = file_timestamp(g)
+    return [(t, ex) for ex in examples]
+
+
+def file_timestamp(filename):
+    return int(os.path.basename(filename).split('-')[0])
+
+
+def _ts_to_str(timestamp):
+    return dt.datetime.fromtimestamp(timestamp).strftime("%Y-%m-%d %H:%M:%S")
+
+
+class ExampleBuffer():
+    def __init__(self, max_size=2000000):
+        self.examples = []
+        self.max_size = max_size
+
+    def parallel_fill(self, games, threads=8, samples_per_game=4):
+        games.sort(key=lambda f: os.path.basename(f))
+        if len(games) * samples_per_game > self.max_size:
+            games = games[-1 * self.max_size // samples_per_game:]
+
+        f = functools.partial(choose, samples_per_game=samples_per_game)
+
+        with mp.Pool(threads) as p:
+             r = tqdm(p.imap(f, games), total=len(games))
+             self.examples = list(itertools.chain(*r))
+
+    def update(self, new_games, samples_per_game=4):
+        """
+        new_games is list of .tfrecord.zz files of new games
+        """
+        new_games.sort(key=lambda f: os.path.basename(f))
+        first_new_game = None
+        for i, game in enumerate(tqdm(new_games)):
+            t = file_timestamp(game)
+            if t <= self.examples[-1][0]:
+                continue
+            elif first_new_game is None:
+                first_new_game = i
+                print("Found {}/{} new games".format(len(new_games)-i, len(new_games)))
+
+            choices = [(t, ex) for ex in pick_examples_from_tfrecord(
+                game, samples_per_game)]
+            if self.count > self.max_size:
+                self.examples = self.examples[samples_per_game:]
```

Comment:
This might be really slow, imagine with 5000 new games each one pushing 4 samples out of this 2000000 long list yikes!

Reply:
moved to deque